### ISSUE TITLE
Filesystem has() call throwing FileNotFoundException

### ISSUE THREAD

**Issue by MichaelMackus** [thephpleague/flysystem-aws-s3-v3#46]
When I call `has` for a directory path, it returns a FileNotFoundException. Calling `has` on the AwsAdapter work correctly.

It looks like the `Filesystem::has` function is stripping the trailing slash from paths, before it calls the `AwsAdapter::has` function. So, AWS doesn't recognize the path since it comes in as a directory path without a trailing slash.

**Comment by frankdejonge** [comment]
@MichaelMackus has calls only concern about files, directories are second class citizens in Flysystem, they're always implicitly created, therefor it's never needed to check for its existence.

**Comment by frankdejonge** [comment]
@MichaelMackus also, which version of the adapter are you using? A recent version uses doesObjectExist rather than fetching metadata, so that exception can not be thrown anymore.

**Comment by MichaelMackus** [comment]
Hmm that's unfortunate, since there are cases in my current project where I have to check existence of directories (its very hierarchical & permissions based).

My composer.lock is locked against this commit: https://github.com/thephpleague/flysystem-aws-s3-v3/commits/e347faaca13ec15bfa3e2dbe02bd7f1ab8e85f55

**Comment by frankdejonge** [comment]
@MichaelMackus there's a new version available, 1.0.6, just released, addressing this issue. An alternative solution would be to list files in that directory, see if it's empty or not?

**Comment by MichaelMackus** [comment]
Thanks much for the help and responsiveness

Updating composer now :)

**Comment by frankdejonge** [comment]
No worries, I'm going to bed now. If the new version or the new approach I suggested fixed your problem, could you close the issue? Otherwise I'll check back here tomorrow.

**Comment by MichaelMackus** [comment]
Perfect, thanks!

**Comment by MichaelMackus** [comment]
Hmm it seems like this issue is actually still persisting.

```$filesystem->has('directory/')```

Always seems to return false :(

It would be nice to have functional methods on directories, since nowhere does it say that "has" on a directory is unsupported.

For now, I'll use listContents as an alternative but it would be nice to have this working.

**Comment by frankdejonge** [comment]
@MichaelMackus this is expected, as I mentioned in the comment earlier, flysystem cares less about directories. have you tried altering your strategy like I suggested with checking for a non-empty listing?

**Comment by MichaelMackus** [comment]
I've worked around this using

```php
if ($filesystem->getAdapter()->has('dir/')) {
  $filesystem->getAdapter()->getMetadata('dir/');
}
```

Kind of messy IMO when this would be much clearer (and avoid working at a low level):

```php
$filesystem->getMetadata('dir/')
```

Unfortunately since Filesystem strips the "/" this doesn't work correctly. It would also be nice for it not to matter whether the "/" is there or not, but these inconsistencies would be better documented/deprecated at least.

**Comment by frankdejonge** [comment]
@MichaelMackus the implementation of that call is exactly what I suggested before:

```php
$isDirectory = ! empty($filesystem->listContents($dirname, false));
```

**Comment by MichaelMackus** [comment]
Hmm that doesn't quite work though. If I do listContents on a valid file path (not a directory), $isDirectory would be true.

**Comment by frankdejonge** [comment]
You mean a file with a trailing `/`? Because listContents ensures the `/` is there.

**Comment by MichaelMackus** [comment]
Yes, it returns an empty array `[]` for a valid file path (non-directory).

**Comment by afroewis** [comment]
Hi! Although the last comment to this issue was back in 2015, I ran into a similar problem today. 

```php
$fs->listContents("assets://some-directory"); // Returns an array with directories and files
$fs->getMetadata("assets://some-directory"); // Throws an exception
```

So `getMetadata` on **any** directory throws an Exception in `AwsS3Adapter`. The same code works perfectly in the `Local` Adapter and does not throw an Exception... Is this really a wanted behavior? 

Do you have a hint how I could check if a node is a file or directory using the `AwsS3Adapter`? `listContents` is not reliable since there could be empty directories.
